File: src-tauri/src/server/peers.rs
use serde::{Deserialize, Serialize};
use std::collections::HashMap;
use std::path::{Path, PathBuf};
use std::sync::Arc;
use std::time::SystemTime;
use tokio::sync::RwLock;
use uuid::Uuid;

use crate::server::log_to_file;

const PEER_CONFIG_FILE: &str = ".org-viewer-peers.json";
const POLL_INTERVAL_SECS: u64 = 30;
const BACKOFF_INTERVAL_SECS: u64 = 120;
const FAILURE_THRESHOLD: u32 = 3;
const HELLO_TIMEOUT_SECS: u64 = 3;

// --- Config types ---

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeerConfig {
    #[serde(rename = "self")]
    pub self_info: PeerSelf,
    pub peers: Vec<PeerEntry>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeerSelf {
    #[serde(rename = "instanceId")]
    pub instance_id: String,
    #[serde(rename = "displayName")]
    pub display_name: String,
    #[serde(rename = "sharedFolders")]
    pub shared_folders: Vec<String>,
    #[serde(rename = "sharedTags")]
    pub shared_tags: Vec<String>,
}

#[derive(Debug, Clone, Serialize, Deserialize)]
pub struct PeerEntry {
    pub name: String,
    pub host: String,
    pub port: u16,
    pub protocol: String,
}

// --- Live status ---

#[derive(Debug, Clone, Serialize)]
pub struct PeerLiveStatus {
    pub name: String,
    pub host: String,
    pub port: u16,
    pub protocol: String,
    pub status: String, // "online" | "offline" | "unknown"
    #[serde(rename = "instanceId", skip_serializing_if = "Option::is_none")]
    pub instance_id: Option<String>,
    #[serde(rename = "displayName", skip_serializing_if = "Option::is_none")]
    pub display_name: Option<String>,
    #[serde(rename = "sharedFolders", skip_serializing_if = "Option::is_none")]
    pub shared_folders: Option<Vec<String>>,
    #[serde(rename = "sharedTags", skip_serializing_if = "Option::is_none")]
    pub shared_tags: Option<Vec<String>>,
    #[serde(rename = "documentCount", skip_serializing_if = "Option::is_none")]
    pub document_count: Option<usize>,
    #[serde(rename = "lastSeen", skip_serializing_if = "Option::is_none")]
    pub last_seen: Option<String>,
    #[serde(rename = "latencyMs", skip_serializing_if = "Option::is_none")]
    pub latency_ms: Option<u64>,
    #[serde(rename = "consecutiveFailures")]
    pub consecutive_failures: u32,
}

// --- Hello response (from remote peer) ---

#[derive(Debug, Deserialize)]
pub struct PeerHelloResponse {
    #[serde(rename = "instanceId")]
    pub instance_id: String,
    #[serde(rename = "displayName")]
    pub display_name: String,
    #[serde(rename = "sharedFolders")]
    pub shared_folders: Vec<String>,
    #[serde(rename = "sharedTags")]
    pub shared_tags: Vec<String>,
    pub stats: PeerHelloStats,
}

#[derive(Debug, Deserialize)]
pub struct PeerHelloStats {
    #[serde(rename = "documentCount")]
    pub document_count: usize,
}

// --- PeerRegistry ---

pub struct PeerRegistry {
    config_path: PathBuf,
    config: RwLock<PeerConfig>,
    status: RwLock<HashMap<String, PeerLiveStatus>>,
    last_config_mtime: RwLock<u64>,
}

impl PeerRegistry {
    pub fn new(org_root: &Path) -> Self {
        let config_path = org_root.join(PEER_CONFIG_FILE);
        let config = Self::load_or_create(&config_path);
        let status = Self::init_status(&config);

        PeerRegistry {
            config_path,
            config: RwLock::new(config),
            status: RwLock::new(status),
            last_config_mtime: RwLock::new(0),
        }
    }

    fn load_or_create(path: &Path) -> PeerConfig {
        if path.exists() {
            if let Ok(raw) = std::fs::read_to_string(path) {
                if let Ok(config) = serde_json::from_str::<PeerConfig>(&raw) {
                    return config;
                }
                log_to_file(&format!("Failed to parse {}", PEER_CONFIG_FILE));
            }
        }

        let config = PeerConfig {
            self_info: PeerSelf {
                instance_id: Uuid::new_v4().to_string(),
                display_name: "My Org".to_string(),
                shared_folders: vec!["knowledge/".to_string()],
                shared_tags: vec![],
            },
            peers: vec![],
        };

        if let Ok(json) = serde_json::to_string_pretty(&config) {
            let _ = std::fs::write(path, json);
            log_to_file(&format!(
                "Created {} with instanceId: {}",
                PEER_CONFIG_FILE, config.self_info.instance_id
            ));
        }

        config
    }

    fn init_status(config: &PeerConfig) -> HashMap<String, PeerLiveStatus> {
        let mut map = HashMap::new();
        for peer in &config.peers {
            let key = format!("{}:{}", peer.host, peer.port);
            map.insert(
                key,
                PeerLiveStatus {
                    name: peer.name.clone(),
                    host: peer.host.clone(),
                    port: peer.port,
                    protocol: peer.protocol.clone(),
                    status: "unknown".to_string(),
                    instance_id: None,
                    display_name: None,
                    shared_folders: None,
                    shared_tags: None,
                    document_count: None,
                    last_seen: None,
                    latency_ms: None,
                    consecutive_failures: 0,
                },
            );
        }
        map
    }

    pub async fn get_self(&self) -> PeerSelf {
        self.config.read().await.self_info.clone()
    }

    pub async fn get_peers(&self) -> Vec<PeerEntry> {
        self.config.read().await.peers.clone()
    }

    pub async fn get_peer_status(&self) -> Vec<PeerLiveStatus> {
        self.status.read().await.values().cloned().collect()
    }

    pub async fn get_online_peers(&self) -> Vec<PeerLiveStatus> {
        self.status
            .read()
            .await
            .values()
            .filter(|p| p.status == "online")
            .cloned()
            .collect()
    }

    /// Start background peer polling task. Returns a JoinHandle.
    pub fn start_polling(self: &Arc<Self>) -> tokio::task::JoinHandle<()> {
        let registry = Arc::clone(self);
        tokio::spawn(async move {
            // Initial poll
            registry.poll_all_peers().await;

            let mut interval =
                tokio::time::interval(std::time::Duration::from_secs(POLL_INTERVAL_SECS));
            loop {
                interval.tick().await;
                registry.poll_all_peers().await;
            }
        })
    }

    async fn poll_all_peers(&self) {
        self.check_config_reload().await;

        let peers = self.config.read().await.peers.clone();
        let mut handles = Vec::new();

        for peer in peers {
            let key = format!("{}:{}", peer.host, peer.port);

            // Check backoff
            let should_skip = {
                let status = self.status.read().await;
                if let Some(s) = status.get(&key) {
                    if s.consecutive_failures >= FAILURE_THRESHOLD {
                        if let Some(last) = &s.last_seen {
                            if let Ok(last_time) = chrono::DateTime::parse_from_rfc3339(last) {
                                let elapsed = chrono::Utc::now()
                                    .signed_duration_since(last_time)
                                    .num_seconds();
                                elapsed < BACKOFF_INTERVAL_SECS as i64
                            } else {
                                false
                            }
                        } else {
                            false
                        }
                    } else {
                        false
                    }
                } else {
                    true // no status entry, skip
                }
            };

            if should_skip {
                continue;
            }

            handles.push(self.poll_peer(peer));
        }

        futures::future::join_all(handles).await;
    }

    async fn poll_peer(&self, peer: PeerEntry) {
        let key = format!("{}:{}", peer.host, peer.port);
        let url = format!(
            "{}://{}:{}/api/federation/hello",
            peer.protocol, peer.host, peer.port
        );

        let start = std::time::Instant::now();

        let client = reqwest::Client::builder()
            .timeout(std::time::Duration::from_secs(HELLO_TIMEOUT_SECS))
            .danger_accept_invalid_certs(true)
            .build()
            .unwrap_or_default();

        match client.get(&url).send().await {
            Ok(resp) if resp.status().is_success() => {
                if let Ok(data) = resp.json::<PeerHelloResponse>().await {
                    let latency = start.elapsed().as_millis() as u64;
                    let mut status = self.status.write().await;
                    if let Some(s) = status.get_mut(&key) {
                        let was_offline = s.status != "online";
                        s.status = "online".to_string();
                        s.instance_id = Some(data.instance_id);
                        s.display_name = Some(data.display_name);
                        s.shared_folders = Some(data.shared_folders);
                        s.shared_tags = Some(data.shared_tags);
                        s.document_count = Some(data.stats.document_count);
                        s.last_seen = Some(chrono::Utc::now().to_rfc3339());
                        s.latency_ms = Some(latency);
                        s.consecutive_failures = 0;

                        if was_offline {
                            log_to_file(&format!("Peer {} ({}): online", peer.name, key));
                        }
                    }
                }
            }
            _ => {
                let mut status = self.status.write().await;
                if let Some(s) = status.get_mut(&key) {
                    let was_online = s.status == "online";
                    s.consecutive_failures += 1;
                    s.status = "offline".to_string();

                    if was_online {
                        log_to_file(&format!("Peer {} ({}): offline", peer.name, key));
                    }
                }
            }
        }
    }

    async fn check_config_reload(&self) {
        let mtime = std::fs::metadata(&self.config_path)
            .ok()
            .and_then(|m| m.modified().ok())
            .and_then(|t| t.duration_since(SystemTime::UNIX_EPOCH).ok())
            .map(|d| d.as_secs())
            .unwrap_or(0);

        let mut last = self.last_config_mtime.write().await;
        if mtime > *last {
            if *last > 0 {
                // Config changed — reload
                let new_config = Self::load_or_create(&self.config_path);
                let old_count = self.config.read().await.peers.len();
                let new_count = new_config.peers.len();

                // Reconcile status map
                let new_keys: std::collections::HashSet<String> = new_config
                    .peers
                    .iter()
                    .map(|p| format!("{}:{}", p.host, p.port))
                    .collect();

                {
                    let mut status = self.status.write().await;
                    // Add new peers
                    for peer in &new_config.peers {
                        let key = format!("{}:{}", peer.host, peer.port);
                        if !status.contains_key(&key) {
                            status.insert(
                                key,
                                PeerLiveStatus {
                                    name: peer.name.clone(),
                                    host: peer.host.clone(),
                                    port: peer.port,
                                    protocol: peer.protocol.clone(),
                                    status: "unknown".to_string(),
                                    instance_id: None,
                                    display_name: None,
                                    shared_folders: None,
                                    shared_tags: None,
                                    document_count: None,
                                    last_seen: None,
                                    latency_ms: None,
                                    consecutive_failures: 0,
                                },
                            );
                        }
                    }
                    // Remove peers no longer in config
                    status.retain(|k, _| new_keys.contains(k));
                }

                *self.config.write().await = new_config;

                if old_count != new_count {
                    log_to_file(&format!(
                        "Peer config hot-reloaded: {} → {} peers",
                        old_count, new_count
                    ));
                }
            }
            *last = mtime;
        }
    }
}

File: packages/server/src/services/peers.ts
import { readFileSync, writeFileSync, existsSync } from 'fs';
import { join } from 'path';
import { randomUUID } from 'crypto';
import type { PeerConfig, PeerEntry, PeerLiveStatus, PeerHelloResponse } from '../types/federation';

const PEER_CONFIG_FILE = '.vitrum-peers.json';
const POLL_INTERVAL_MS = 30_000;
const BACKOFF_INTERVAL_MS = 120_000;
const FAILURE_THRESHOLD = 3;
const HELLO_TIMEOUT_MS = 3_000;

export class PeerRegistry {
  private config: PeerConfig;
  private configPath: string;
  private peerStatus: Map<string, PeerLiveStatus> = new Map();
  private pollTimer: ReturnType<typeof setInterval> | null = null;
  private onStatusChange?: (peer: PeerLiveStatus) => void;

  constructor(orgRoot: string) {
    this.configPath = join(orgRoot, PEER_CONFIG_FILE);
    this.config = this.loadOrCreate();
    this.initPeerStatus();
  }

  private loadOrCreate(): PeerConfig {
    if (existsSync(this.configPath)) {
      try {
        const raw = readFileSync(this.configPath, 'utf-8');
        return JSON.parse(raw);
      } catch (e) {
        console.error(`Failed to parse ${PEER_CONFIG_FILE}:`, e);
      }
    }

    // Create default config
    const config: PeerConfig = {
      self: {
        instanceId: randomUUID(),
        displayName: 'My Org',
        sharedFolders: ['knowledge/'],
        sharedTags: [],
      },
      peers: [],
    };

    this.save(config);
    console.log(`Created ${PEER_CONFIG_FILE} with instanceId: ${config.self.instanceId}`);
    return config;
  }

  private save(config: PeerConfig): void {
    writeFileSync(this.configPath, JSON.stringify(config, null, 2), 'utf-8');
  }

  private initPeerStatus(): void {
    for (const peer of this.config.peers) {
      const key = `${peer.host}:${peer.port}`;
      this.peerStatus.set(key, {
        name: peer.name,
        host: peer.host,
        port: peer.port,
        protocol: peer.protocol,
        status: 'unknown',
        consecutiveFailures: 0,
      });
    }
  }

  getSelf() {
    return this.config.self;
  }

  getPeers(): PeerEntry[] {
    return this.config.peers;
  }

  getPeerStatus(): PeerLiveStatus[] {
    return Array.from(this.peerStatus.values());
  }

  getOnlinePeers(): PeerLiveStatus[] {
    return this.getPeerStatus().filter(p => p.status === 'online');
  }

  isPeerRegistered(host: string): boolean {
    return this.config.peers.some(
      p => p.host === host || p.host.split('.')[0] === host.split('.')[0]
    );
  }

  getPeerByHost(host: string): PeerEntry | undefined {
    return this.config.peers.find(
      p => p.host === host || p.host.split('.')[0] === host.split('.')[0]
    );
  }

  onPeerStatusChange(callback: (peer: PeerLiveStatus) => void) {
    this.onStatusChange = callback;
  }

  startPolling(): void {
    if (this.pollTimer) return;

    // Initial poll
    this.pollAllPeers();

    // Set up interval
    this.pollTimer = setInterval(() => {
      this.pollAllPeers();
    }, POLL_INTERVAL_MS);

    console.log(`Peer polling started (${POLL_INTERVAL_MS / 1000}s interval, ${this.config.peers.length} peers)`);
  }

  stopPolling(): void {
    if (this.pollTimer) {
      clearInterval(this.pollTimer);
      this.pollTimer = null;
    }
  }

  private async pollAllPeers(): Promise<void> {
    // Hot-reload config if changed on disk
    this.checkConfigReload();

    const promises = this.config.peers.map(peer => this.pollPeer(peer));
    await Promise.allSettled(promises);
  }

  private lastConfigMtime = 0;

  private checkConfigReload(): void {
    try {
      const file = Bun.file(this.configPath);
      // Use sync file read to check mtime
      const stat = require('fs').statSync(this.configPath);
      const mtime = stat.mtimeMs;
      if (mtime > this.lastConfigMtime) {
        if (this.lastConfigMtime > 0) {
          // Config changed — reload
          const oldPeerCount = this.config.peers.length;
          this.config = this.loadOrCreate();

          // Reconcile peer status map: add new peers, keep existing status for known peers
          const currentKeys = new Set(this.peerStatus.keys());
          const newKeys = new Set(this.config.peers.map(p => `${p.host}:${p.port}`));

          // Add new peers
          for (const peer of this.config.peers) {
            const key = `${peer.host}:${peer.port}`;
            if (!this.peerStatus.has(key)) {
              this.peerStatus.set(key, {
                name: peer.name,
                host: peer.host,
                port: peer.port,
                protocol: peer.protocol,
                status: 'unknown',
                consecutiveFailures: 0,
              });
            }
          }

          // Remove peers no longer in config
          for (const key of currentKeys) {
            if (!newKeys.has(key)) {
              this.peerStatus.delete(key);
            }
          }

          if (this.config.peers.length !== oldPeerCount) {
            console.log(`Peer config hot-reloaded: ${oldPeerCount} → ${this.config.peers.length} peers`);
          }
        }
        this.lastConfigMtime = mtime;
      }
    } catch {
      // Config file might not exist yet, ignore
    }
  }

  private async pollPeer(peer: PeerEntry): Promise<void> {
    const key = `${peer.host}:${peer.port}`;
    const status = this.peerStatus.get(key);
    if (!status) return;

    // Skip if in backoff
    if (status.consecutiveFailures >= FAILURE_THRESHOLD) {
      // Only poll at backoff interval
      const lastSeen = status.lastSeen ? new Date(status.lastSeen).getTime() : 0;
      if (Date.now() - lastSeen < BACKOFF_INTERVAL_MS) {
        return;
      }
    }

    const url = `${peer.protocol}://${peer.host}:${peer.port}/api/federation/hello`;
    const startMs = Date.now();

    try {
      const controller = new AbortController();
      const timeout = setTimeout(() => controller.abort(), HELLO_TIMEOUT_MS);

      const response = await fetch(url, {
        signal: controller.signal,
        headers: { 'Accept': 'application/json' },
      });
      clearTimeout(timeout);

      if (!response.ok) {
        throw new Error(`HTTP ${response.status}`);
      }

      const data = await response.json() as PeerHelloResponse;
      const latencyMs = Date.now() - startMs;
      const oldStatus = status.status;

      status.status = 'online';
      status.instanceId = data.instanceId;
      status.displayName = data.displayName;
      status.sharedFolders = data.sharedFolders;
      status.sharedTags = data.sharedTags;
      status.documentCount = data.stats.documentCount;
      status.lastSeen = new Date().toISOString();
      status.latencyMs = latencyMs;
      status.consecutiveFailures = 0;

      if (oldStatus !== 'online' && this.onStatusChange) {
        this.onStatusChange(status);
      }
    } catch (e) {
      const oldStatus = status.status;
      status.consecutiveFailures++;
      status.status = 'offline';

      if (oldStatus === 'online' && this.onStatusChange) {
        this.onStatusChange(status);
      }
    }
  }

  // Reload config from disk (for hot-reload when user edits config)
  reload(): void {
    this.config = this.loadOrCreate();
    this.initPeerStatus();
    console.log(`Peer config reloaded: ${this.config.peers.length} peers`);
  }
}
